<a href="https://colab.research.google.com/github/jenyadev3/Telecom-customer-base-segmentation/blob/main/vkr_telecom.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Сегментация клиентской базы телекоммуникационной компании

**Аналитическая задача** — провести анализ данных с целью выделения наиболее типичных групп клиентов и разработки предложений для каждой из групп.

## Описание данных
Каждый клиент описывается следующим набором признаков:
- `Возраст`, `Среднемесячный расход`, `Средняя продолжительность разговоров`, `Звонков днем за месяц`, `Звонков вечером за месяц`, `Звонков ночью за месяц`, `Звонки в другие города`, `Звонки в другие страны`, `Доля звонков на стационарные телефоны`, `Количество SMS за месяц`, `Дата подключения тарифа`.

## Примерный план по выполнению проекта

**Шаг 1.** Загрузка данных;

**Шаг 2.** Первичная обработка данных (при необходимости):
- скорректировать заголовки;
- скорректировать типы признаков;
- проверить наличие дублирующихся записей;
- проверить наличие аномальных значений;
- восстановить пропущенные значения;

**Шаг 3.** Добавление новых переменных:
- по значениям признака `Возраст` введите новую переменную `Возрастная категория`, принимающую значения "студент", "аспирант", "бизнесмен" и "знаток" по следующую правилу:

$$
\text{Возрастная категория} =
\left\{
\begin{array}{l}
\text{студент,} \quad \text{если Возраст } \in [19, 24];\\  
\text{аспирант,} \quad \text{если Возраст } \in [25, 33];\\
\text{бизнесмен,} \quad \text{если Возраст } \in [34, 56];\\
\text{знаток,} \quad \text{если Возраст } \in [57, 70].
\end{array}
\right.
$$

- по значениям признака `Дата подключения тарифа` создайте признаки: `Год подключения`, `Месяц подключения`, `Дата подключения`;  

**Шаг 3.** Провести исследовательский анализ данных:
- в разрезе значений признаков `Год подключения`, `Месяц подключения`, `Дата подключения` исследуйте:
    - динамику подключения к тарифам (количество клиентов). Постройте графики. В какой год, месяц и день подключались меньше/больше всего клиентов? Какие выводы можно сделать?
- в разрезе значений признака `Возрастная категория` исследуйте распределение признаков `Среднемесячный расход`, `Средняя продолжительность разговоров`, `Звонков днем за месяц`, `Звонков вечером за месяц`, `Звонков ночью за месяц`, `Звонки в другие города`, `Доля звонков на стационарные телефоны`, `Количество SMS за месяц`. Для каждого из признаков рассчитайте выборочное среднее, медиану и моду. Постройте графики. Какие выводы можно сделать о предпочтениях клиентов разных возрастных категорий в отношении используемых услуг (звонков и SMS; времени суток);
- клиенты каких возрастных категорий (ТОП-2):
  
    - больше всего в среднем в месяц тратят на оплату услуг связи;
    - больше всего тратят времени на общение в месяц днем, вечером и ночью;
    - больше всего по количеству звонков в месяц днем, вечером и ночью. Совпадают ли результаты с предыдущем пунктом;

- с помощью диаграмм рассеивания исследуейте зависимости между признаками `Среднемесячный расход`, `Средняя продолжительность разговоров`, `Звонков днем за месяц`, `Звонков вечером за месяц`, `Звонков ночью за месяц`, `Звонки в другие города`, `Доля звонков на стационарные телефоны`, `Количество SMS за месяц`. Какие выводы можно сделать?

**1. Загрузка данных**

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("/content/dataset_telecom.csv")

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4492 entries, 0 to 4491
Data columns (total 11 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Возраст                                4492 non-null   int64  
 1   Среднемесячный расход                  4468 non-null   float64
 2   Средняя продолжительность разговоров   4475 non-null   float64
 3   Звонков днем за месяц                  4472 non-null   float64
 4   Звонков вечером за месяц               4489 non-null   float64
 5   Звонков ночью за месяц                 4492 non-null   object 
 6   Звонки в другие города                 4492 non-null   object 
 7   Звонки в другие страны                 4492 non-null   int64  
 8   Доля звонков на стационарные телефоны  4492 non-null   object 
 9   Количество SMS за месяц                4492 non-null   object 
 10  Дата подключения тарифа                4492 non-null   object 
dtypes: f

In [ ]:
df.describe()

Возраст  Среднемесячный расход  \
count  4492.000000            4468.000000   
mean     41.894924             505.533577   
std      13.077254             646.354693   
min      19.000000               3.180000   
25%      31.000000             152.490000   
50%      43.000000             315.510000   
75%      52.000000             599.840000   
max      70.000000            5142.760000   

       Средняя продолжительность разговоров  Звонков днем за месяц  \
count                           4475.000000            4472.000000   
mean                               4.230827              63.899374   
std                                3.005210              62.877394   
min                                0.100000               1.000000   
25%                                2.100000              37.000000   
50%                                3.300000              53.000000   
75%                                5.900000              68.000000   
max                               20.000000             500.000000   

       Звонков вечером за месяц  Звонки в другие страны  
count               4489.000000             4492.000000  
mean                  70.363778                0.394034  
std                   41.219909                1.171980  
min                    1.000000                0.000000  
25%                   42.000000                0.000000  
50%                   71.000000                0.000000  
75%                   98.000000                0.000000  
max                  160.000000               12.000000

Первичный осмотр выявил, что в дата фрейме всего 10 столбцов и 4492 записей. В некоторых столбцах не хватает данных. Также требуется смена типов у некоторых столбцов.
Это был первичный анализ, обработки будут в следующих шагах.

**Шаг 2. Первичная обработка данных**

*Скорректировать типы признаков*

In [ ]:
df.info() #смотрю какие типы признаков вообще

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4492 entries, 0 to 4491
Data columns (total 11 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   Возраст                                4492 non-null   int64  
 1   Среднемесячный расход                  4468 non-null   float64
 2   Средняя продолжительность разговоров   4475 non-null   float64
 3   Звонков днем за месяц                  4472 non-null   float64
 4   Звонков вечером за месяц               4489 non-null   float64
 5   Звонков ночью за месяц                 4492 non-null   object 
 6   Звонки в другие города                 4492 non-null   object 
 7   Звонки в другие страны                 4492 non-null   int64  
 8   Доля звонков на стационарные телефоны  4492 non-null   object 
 9   Количество SMS за месяц                4492 non-null   object 
 10  Дата подключения тарифа                4492 non-null   object 
dtypes: f

In [ ]:
#перевожу время из object
df["Дата подключения тарифа"] = pd.to_datetime(df["Дата подключения тарифа"],errors = "coerce")

In [ ]:
#перевела из object в числа
cols = ["Звонков ночью за месяц","Звонки в другие города","Доля звонков на стационарные телефоны","Количество SMS за месяц"]
for c in cols:
    df[c] = pd.to_numeric(df[c], errors="coerce")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4492 entries, 0 to 4491
Data columns (total 11 columns):
 #   Column                                 Non-Null Count  Dtype         
---  ------                                 --------------  -----         
 0   Возраст                                4492 non-null   int64         
 1   Среднемесячный расход                  4468 non-null   float64       
 2   Средняя продолжительность разговоров   4475 non-null   float64       
 3   Звонков днем за месяц                  4472 non-null   float64       
 4   Звонков вечером за месяц               4489 non-null   float64       
 5   Звонков ночью за месяц                 4491 non-null   float64       
 6   Звонки в другие города                 4490 non-null   float64       
 7   Звонки в другие страны                 4492 non-null   int64         
 8   Доля звонков на стационарные телефоны  4491 non-null   float64       
 9   Количество SMS за месяц                4491 non-null   float64 

*Проверить наличие дублирующихся записей*

In [ ]:
df.duplicated().sum() #показывает ноль, значит дубликатов нет

np.int64(0)

*Проверить наличие аномальных значений*

*Отрицательные значения*

In [ ]:
#вычисляем то что реально числовые колонки

num_cols = df.select_dtypes(include=[np.number]).columns
num_cols


Index(['Возраст', 'Среднемесячный расход',
       'Средняя продолжительность разговоров', 'Звонков днем за месяц',
       'Звонков вечером за месяц', 'Звонков ночью за месяц',
       'Звонки в другие города', 'Звонки в другие страны',
       'Доля звонков на стационарные телефоны', 'Количество SMS за месяц'],
      dtype='object')

In [ ]:
(df[num_cols] < 0).any().any()

np.False_

*Максимальные и минимальные значение*

In [ ]:
df.describe()

Возраст  Среднемесячный расход  \
count  4492.000000            4468.000000   
mean     41.894924             505.533577   
min      19.000000               3.180000   
25%      31.000000             152.490000   
50%      43.000000             315.510000   
75%      52.000000             599.840000   
max      70.000000            5142.760000   
std      13.077254             646.354693   

       Средняя продолжительность разговоров  Звонков днем за месяц  \
count                           4475.000000            4472.000000   
mean                               4.230827              63.899374   
min                                0.100000               1.000000   
25%                                2.100000              37.000000   
50%                                3.300000              53.000000   
75%                                5.900000              68.000000   
max                               20.000000             500.000000   
std                                3.005210              62.877394   

       Звонков вечером за месяц  Звонков ночью за месяц  \
count               4489.000000             4491.000000   
mean                  70.363778                6.272990   
min                    1.000000                0.000000   
25%                   42.000000                0.000000   
50%                   71.000000                1.000000   
75%                   98.000000                5.000000   
max                  160.000000              110.000000   
std                   41.219909               14.714668   

       Звонки в другие города  Звонки в другие страны  \
count             4490.000000             4492.000000   
mean                 8.658575                0.394034   
min                  0.000000                0.000000   
25%                  0.000000                0.000000   
50%                  2.000000                0.000000   
75%                 12.000000                0.000000   
max                 88.000000               12.000000   
std                 13.477005                1.171980   

       Доля звонков на стационарные телефоны  Количество SMS за месяц  \
count                            4491.000000              4491.000000   
mean                               10.333556                21.245380   
min                                 0.000000                 0.000000   
25%                                 3.000000                 3.000000   
50%                                 8.000000                 6.000000   
75%                                16.000000                32.000000   
max                                60.000000               179.000000   
std                                 9.222882                27.914631   

             Дата подключения тарифа  
count                           4492  
mean   2018-07-10 01:04:50.785841664  
min              2015-01-01 15:48:33  
25%       2016-10-21 08:23:49.500000  
50%       2018-06-21 16:56:24.500000  
75%    2020-04-04 18:48:26.750000128  
max              2021-12-31 14:12:11  
std                              NaN

*Восстановить пропущенные значения*

In [ ]:
df.isna().any()

Возраст                                  False
Среднемесячный расход                     True
Средняя продолжительность разговоров      True
Звонков днем за месяц                     True
Звонков вечером за месяц                  True
Звонков ночью за месяц                    True
Звонки в другие города                    True
Звонки в другие страны                   False
Доля звонков на стационарные телефоны     True
Количество SMS за месяц                   True
Дата подключения тарифа                  False
dtype: bool

In [ ]:
df.fillna(0)

Возраст  Среднемесячный расход  Средняя продолжительность разговоров  \
0          24                   0.00                                   2.4   
1          51                 287.51                                   1.7   
2          41                 113.70                                   2.1   
3          35                 410.23                                   5.6   
4          26                 537.60                                   4.8   
...       ...                    ...                                   ...   
4487       40                 186.20                                   3.7   
4488       40                 500.68                                   6.0   
4489       60                 470.42                                   3.5   
4490       38                 858.99                                   9.3   
4491       33                 151.92                                   1.5   

      Звонков днем за месяц  Звонков вечером за месяц  Звонков ночью за месяц  \
0                      12.0                      65.0                     5.0   
1                     111.0                     109.0                     1.0   
2                      41.0                      27.0                     0.0   
3                      47.0                      49.0                     0.0   
4                      58.0                      77.0                     4.0   
...                     ...                       ...                     ...   
4487                   36.0                      28.0                     2.0   
4488                   46.0                      66.0                     0.0   
4489                   58.0                     129.0                     2.0   
4490                   50.0                      74.0                     0.0   
4491                   59.0                      75.0                     2.0   

      Звонки в другие города  Звонки в другие страны  \
0                        0.0                       0   
1                       44.0                       0   
2                        0.0                       0   
3                        0.0                       0   
4                        0.0                       0   
...                      ...                     ...   
4487                     6.0                       0   
4488                     6.0                       0   
4489                     4.0                       0   
4490                     0.0                       0   
4491                     1.0                       0   

      Доля звонков на стационарные телефоны  Количество SMS за месяц  \
0                                       5.0                     56.0   
1                                       6.0                      1.0   
2                                       1.0                     36.0   
3                                      11.0                     23.0   
4                                      16.0                     29.0   
...                                     ...                      ...   
4487                                   12.0                     30.0   
4488                                   15.0                     23.0   
4489                                   19.0                      5.0   
4490                                   28.0                     69.0   
4491                                    8.0                     24.0   

     Дата подключения тарифа  
0        2018-06-17 12:14:35  
1        2021-10-21 15:39:54  
2        2015-03-26 11:26:15  
3        2016-01-04 15:53:20  
4        2017-05-03 13:33:53  
...                      ...  
4487     2021-04-10 14:58:23  
4488     2021-09-22 14:52:43  
4489     2019-04-04 17:33:38  
4490     2015-12-30 16:28:26  
4491     2019-04-14 11:59:10  

[4492 rows x 11 columns]

In [ ]:
df.corr()

Возраст  Среднемесячный расход  \
Возраст                                1.000000              -0.015482   
Среднемесячный расход                 -0.015482               1.000000   
Средняя продолжительность разговоров  -0.045908               0.830043   
Звонков днем за месяц                 -0.011976               0.843385   
Звонков вечером за месяц              -0.192496               0.421987   
Звонков ночью за месяц                -0.267727               0.392874   
Звонки в другие города                 0.028714               0.334365   
Звонки в другие страны                 0.024974               0.381036   
Доля звонков на стационарные телефоны  0.120523               0.357391   
Количество SMS за месяц               -0.592415               0.090872   
Дата подключения тарифа               -0.003979              -0.017723   

                                       Средняя продолжительность разговоров  \
Возраст                                                           -0.045908   
Среднемесячный расход                                              0.830043   
Средняя продолжительность разговоров                               1.000000   
Звонков днем за месяц                                              0.516141   
Звонков вечером за месяц                                           0.321616   
Звонков ночью за месяц                                             0.397310   
Звонки в другие города                                             0.260530   
Звонки в другие страны                                             0.230470   
Доля звонков на стационарные телефоны                              0.239661   
Количество SMS за месяц                                            0.183775   
Дата подключения тарифа                                           -0.031687   

                                       Звонков днем за месяц  \
Возраст                                            -0.011976   
Среднемесячный расход                               0.843385   
Средняя продолжительность разговоров                0.516141   
Звонков днем за месяц                               1.000000   
Звонков вечером за месяц                            0.314136   
Звонков ночью за месяц                              0.317474   
Звонки в другие города                              0.369247   
Звонки в другие страны                              0.441975   
Доля звонков на стационарные телефоны               0.418755   
Количество SMS за месяц                             0.028308   
Дата подключения тарифа                            -0.010504   

                                       Звонков вечером за месяц  \
Возраст                                               -0.192496   
Среднемесячный расход                                  0.421987   
Средняя продолжительность разговоров                   0.321616   
Звонков днем за месяц                                  0.314136   
Звонков вечером за месяц                               1.000000   
Звонков ночью за месяц                                 0.285896   
Звонки в другие города                                 0.198853   
Звонки в другие страны                                 0.117303   
Доля звонков на стационарные телефоны                  0.200196   
Количество SMS за месяц                                0.235010   
Дата подключения тарифа                               -0.038597   

                                       Звонков ночью за месяц  \
Возраст                                             -0.267727   
Среднемесячный расход                                0.392874   
Средняя продолжительность разговоров                 0.397310   
Звонков днем за месяц                                0.317474   
Звонков вечером за месяц                             0.285896   
Звонков ночью за месяц                               1.000000   
Звонки в другие города                               0.142602   
Звонки в другие страны                               0.125934   
Доля звонков на стационарные телефоны      

**Шаг 3. Провести исследовательский анализ данных:**

В разрезе значений признаков Год подключения, Месяц подключения, Дата подключения будет рассмотренна:
в разрезе значений признаков Год подключения, Месяц подключения, Дата
подключения исследуйте:

динамику подключения к тарифам (количество клиентов).
Постройте графики.
В какой год, месяц и день подключались меньше/больше всего клиентов? Какие выводы можно сделать?
в разрезе значений признака
*динамика подключения к тарифам (количество клиентов)*

In [ ]:
df_sorted = df.sort_values(by='Возраст', ascending=False)
df_sorted

Возраст  Среднемесячный расход  Средняя продолжительность разговоров  \
2158       70                  34.68                                   1.7   
421        70                  37.44                                   1.3   
585        70                  15.96                                   1.9   
409        70                  62.64                                   NaN   
3775       70                  17.28                                   1.6   
...       ...                    ...                                   ...   
4463       19                  82.94                                   1.2   
4455       19                 149.15                                   2.1   
3850       19                 552.47                                   5.9   
3876       19                  20.73                                   0.3   
31         19                  52.43                                   0.7   

      Звонков днем за месяц  Звонков вечером за месяц  Звонков ночью за месяц  \
2158                    5.0                      12.0                     0.0   
421                    18.0                       6.0                     0.0   
585                     5.0                       2.0                     0.0   
409                    21.0                       8.0                     0.0   
3775                    7.0                       2.0                     0.0   
...                     ...                       ...                     ...   
4463                   32.0                      74.0                     5.0   
4455                   44.0                      63.0                     9.0   
3850                   44.0                     103.0                     0.0   
3876                   39.0                      68.0                     3.0   
31                     29.0                      74.0                     8.0   

      Звонки в другие города  Звонки в другие страны  \
2158                     0.0                       0   
421                      0.0                       0   
585                      0.0                       0   
409                      0.0                       0   
3775                     0.0                       0   
...                      ...                     ...   
4463                     2.0                       0   
4455                     0.0                       0   
3850                     0.0                       0   
3876                     0.0                       0   
31                       0.0                       0   

      Доля звонков на стационарные телефоны  Количество SMS за месяц  \
2158                                    9.0                      1.0   
421                                     4.0                      3.0   
585                                     5.0                      3.0   
409                                     1.0                      1.0   
3775                                   12.0                      0.0   
...                                     ...                      ...   
4463                                   14.0                     25.0   
4455                                    0.0                     22.0   
3850                                    5.0                     56.0   
3876                                    4.0                     14.0   
31                                      0.0                     39.0   

     Дата подключения тарифа  
2158     2020-04-26 18:49:20  
421      2019-06-11 12:34:50  
585      2020-07-26 11:22:40  
409      2021-06-12 10:09:39  
3775     2021-02-08 11:41:53  
...                      ...  
4463     2018-03-23 10:37:57  
4455     2019-02-12 10:54:40  
3850     2018-10-26 14:34:19  
3876     2021-09-07 15:22:07  
31       2018-07-20 10:59:46  

[4492 rows x 11 columns]

In [ ]:
df["Дата подключения тарифа"].value_counts()
df["Год"] = df["Дата подключения тарифа"].dt.year
df["Месяц"] = df["Дата подключения тарифа"].dt.month
df["День"] = df["Дата подключения тарифа"].dt.day
df.head(20)


Возраст  Среднемесячный расход  Средняя продолжительность разговоров  \
0        24                    NaN                                   2.4   
1        51                 287.51                                   1.7   
2        41                 113.70                                   2.1   
3        35                 410.23                                   5.6   
4        26                 537.60                                   4.8   
5        26                 324.56                                   3.3   
6        50                 410.04                                   4.1   
7        66                  14.52                                   1.1   
8        36                 209.12                                   2.5   
9        40                 162.40                                   1.6   
10       23                2672.34                                  13.8   
11       29                 556.92                                   3.4   
12       43                 636.75                                   NaN   
13       21                  57.33                                   2.0   
14       37                 339.20                                   2.5   
15       69                  50.40                                   2.1   
16       46                 539.28                                   5.1   
17       37                 222.09                                   1.9   
18       48                 202.30                                   3.4   
19       52                  17.69                                   NaN   

    Звонков днем за месяц  Звонков вечером за месяц  Звонков ночью за месяц  \
0                    12.0                      65.0                     5.0   
1                   111.0                     109.0                     1.0   
2                    41.0                      27.0                     0.0   
3                    47.0                      49.0                     0.0   
4                    58.0                      77.0                     4.0   
5                    45.0                      75.0                     3.0   
6                    61.0                      71.0                     0.0   
7                    10.0                       1.0                     0.0   
8                    34.0                      70.0                     3.0   
9                    79.0                      32.0                     4.0   
10                  222.0                      77.0                    10.0   
11                   97.0                      96.0                     2.0   
12                   59.0                      64.0                     0.0   
13                    2.0                      39.0                     8.0   
14                   41.0                     152.0                     6.0   
15                   15.0                       5.0                     0.0   
16                   92.0                      37.0                     0.0   
17                   61.0                      95.0                     5.0   
18                   31.0                      41.0                     0.0   
19                    7.0                       5.0                     0.0   

    Звонки в другие города  Звонки в другие страны  \
0                      0.0                       0   
1                     44.0                       0   
2                      0.0                       0   
3                      0.0                       0   
4                      0.0                       0   
5                      1.0                       1   
6                      0.0                       2   
7                      0.0                       0   
8                      2.0                       0   
9                      0.0                       0   
10                    30.0                       0   
11                     0.0                       0   
12                     0.0                   

клиенты каких возрастных категорий (ТОП-2):

больше всего в среднем в месяц тратят на оплату услуг связи;
больше всего тратят времени на общение в месяц днем, вечером и ночью;
больше всего по количеству звонков в месяц днем, вечером и ночью. Совпадают ли результаты с предыдущем пунктом;
с помощью диаграмм рассеивания исследуейте зависимости между признаками Среднемесячный расход, Средняя продолжительность разговоров, Звонков днем за месяц, Звонков вечером за месяц, Звонков ночью за месяц, Звонки в другие города, Доля звонков на стационарные телефоны, Количество SMS за месяц. Какие выводы можно сделать?